thingsboard_gateway/connectors/modbus_async/modbus_connector.py

```diff
@@ -0,0 +1,697 @@
+                                             port=slave.port,
+                                             timeout=slave.timeout,
+                                             retry_on_empty=slave.retry_on_empty,
+                                             retry_on_invalid=slave.retry_on_invalid,
+                                             retries=slave.retries,
+                                             baudrate=slave.baudrate,
+                                             stopbits=slave.stopbits,
+                                             bytesize=slave.bytesize,
+                                             parity=slave.parity,
+                                             strict=slave.strict)
+        else:
+            raise Exception("Invalid Modbus transport type.")
+
+        return master
+
+    def __get_master_connection(self, slave: Slave):
+        """
+        Method check if connection to master already exists and return it
+        or create new connection and return it
+
+        Newly created master connection structure will look like:
+        self._master_connections = {
+            '127.0.0.1:5021': AsyncClient object
+        }
+        """
+
+        socket_str = slave.host + ':' + str(slave.port)
+        if socket_str not in self._master_connections:
+            master_connection = self.__configure_master(slave)
+            self._master_connections[socket_str] = master_connection
+
+        return self._master_connections[socket_str]
+
+    def __add_slave(self, slave_config):
+        slave_config.update({
+            'connector': self,
+            'gateway': self.__gateway,
+            'logger': self.__log,
+            'callback': AsyncModbusConnector.callback
+        })
+
+        slave = Slave(**slave_config)
+        master_connection = self.__get_master_connection(slave)
+        slave.set_master_connection(master_connection)  # set master connection only via set_master_connection method
+
+        self.__slaves.append(slave)
+
+    def __add_slaves(self, slaves_config):
+        for slave_config in slaves_config:
+            try:
+                self.__add_slave(slave_config)
+            except Exception as e:
+                self.__log.exception('Failed to add slave: %s', e)
+
+        self.__log.debug('Added %d slaves', len(self.__slaves))
+
+    @classmethod
+    def callback(cls, slave: Slave):
+        cls.PROCESS_REQUESTS.put_nowait(slave)
+
+    async def __process_requests(self):
+        while not self.__stopped:
+            try:
+                slave = self.PROCESS_REQUESTS.get_nowait()
+                await self.__poll_device(slave)
+            except Empty:
+                await asyncio.sleep(.01)
+            except Exception as e:
+                self.__log.exception(e)
+
+    async def __poll_device(self, slave):
+        self.__log.debug("Polling %s slave", slave)
+
+        if slave.type == 'serial':
+            await self.__lock.acquire()
```

Comment:
Let’s refactor this point to avoid sending requests to platform inside the connector lock, you can add some flag and send all required requests after blocking operation.